blockchain/upgrade.go

```diff
@@ -3363,6 +3453,449 @@ func separateUtxoDatabase(ctx context.Context, db database.DB, utxoDb database.D
+		var b [8]byte
+		byteOrder.PutUint64(b[:], ui64)
+		return b[:]
+	}
+
+	// Store the database version.
+	verKey := prefixedKey(bucketIdV1, utxoDbInfoVersionKeyNameV1)
+	err = tx.Put(verKey, uint32Bytes(info.version))
+	if err != nil {
+		return err
+	}
+
+	// Store the compression version.
+	compVerKey := prefixedKey(bucketIdV1, utxoDbInfoCompVerKeyNameV1)
+	err = tx.Put(compVerKey, uint32Bytes(info.compVer))
+	if err != nil {
+		return err
+	}
+
+	// Store the UTXO set version.
+	utxoVerKey := prefixedKey(bucketIdV1, utxoDbInfoUtxoVerKeyNameV1)
+	err = tx.Put(utxoVerKey, uint32Bytes(info.utxoVer))
+	if err != nil {
+		return err
+	}
+
+	// Store the database creation date.
+	createdKey := prefixedKey(bucketIdV1, utxoDbInfoCreatedKeyNameV1)
+	return tx.Put(createdKey, uint64Bytes(uint64(info.created.Unix())))
+}
+
+// upgradeUtxoDbToVersion2 upgrades a UTXO database from version 1 to version 2.
+func upgradeUtxoDbToVersion2(ctx context.Context, utxoBackend UtxoBackend) error {
+	if interruptRequested(ctx) {
+		return errInterruptRequested
+	}
+
+	log.Info("Upgrading UTXO database to version 2...")
+	start := time.Now()
+
+	// Migrate the UTXO data to use simple key prefixes rather than buckets.
+	err := migrateUtxoDbBuckets(ctx, utxoBackend)
+	if err != nil {
+		return err
+	}
+
+	// Fetch the backend versioning info.
+	utxoDbInfo, err := utxoBackend.FetchInfo()
+	if err != nil {
+		return err
+	}
+
+	// Update and persist the UTXO database version.
+	utxoDbInfo.version = 2
+	err = utxoBackend.PutInfo(utxoDbInfo)
+	if err != nil {
+		return err
+	}
+
+	// Update and persist the UTXO database version in the legacy V1 bucket.
+	// This allows older versions to identify that a newer database version exists
+	// in the case of a downgrade.
+	err = utxoBackend.Update(func(tx UtxoBackendTx) error {
+		return dbPutUtxoBackendInfoV1(tx, utxoDbInfo)
+	})
+	if err != nil {
+		return err
+	}
+
+	elapsed := time.Since(start).Round(time.Millisecond)
+	log.Infof("Done upgrading database in %v.", elapsed)
+	return nil
+}
+
+// moveUtxoDatabase moves the UTXO database from the provided old path to the
+// provided new path.  The database must exist at the provided old path.
+func moveUtxoDatabase(ctx context.Context, oldPath string, newPath string) error {
+	log.Info("Moving UTXO database.  This may take a while...")
+	start := time.Now()
+
+	// Options for open the leveldb database.
+	opts := opt.Options{
+		Strict:      opt.DefaultStrict,
+		Compression: opt.NoCompression,
+		Filter:      filter.NewBloomFilter(10),
+	}
+
+	// Open the database at the old path.
+	oldDb, err := leveldb.OpenFile(oldPath, &opts)
+	if err != nil {
+		str := fmt.Sprintf("failed to open UTXO database: %v", err)
+		return convertLdbErr(err, str)
+	}
+
+	// Create and open the database at the new path.
+	newDb, err := leveldb.OpenFile(newPath, &opts)
+	if err != nil {
+		oldDb.Close()
+		str := fmt.Sprintf("failed to open UTXO database: %v", err)
```

Comment:
I feel like these two error messages could include the words "old" and "new" unless you feel the `err` message itself would be sufficient to disambiguate.

Reply:
Good suggestion.  Updated the error messages to include "old" and "new".